jaxopt/_src/lbfgs.py

```diff
@@ -38,39 +38,33 @@
 from jaxopt.tree_util import tree_l2_norm
 
 
-def ihp_body_right(q, tup):
-  s, y, rho = tup
-  alpha = rho * tree_vdot(s, q)
-  q = tree_add_scalar_mul(q, -alpha, y)  # q = q - alpha * y
-  return q, alpha
-
-
-def ihp_body_left(r, tup):
-  s, y, rho, alpha = tup
-  beta = rho * tree_vdot(y, r)
-  r = tree_add_scalar_mul(r, alpha - beta, s)  # r = r + (alpha - beta) * s
-  return r, beta
-
-
 def inv_hessian_product_leaf(v: jnp.ndarray,
                              s_history: jnp.ndarray,
                              y_history: jnp.ndarray,
                              rho_history: jnp.ndarray,
-                             gamma: float = 1.0):
+                             gamma: float = 1.0,
+                             start: int = 0):
+
+  history_size = len(s_history)
+
+  indices = (start + jnp.arange(history_size)) % history_size
 
-  # Compute right part.
-  q, alpha = jax.lax.scan(ihp_body_right,
-                          v,
-                          (s_history, y_history, rho_history),
-                          reverse=True)
+  def body_right(r, i):
+    alpha = rho_history[i] * jnp.vdot(s_history[i], r)
+    r = r - alpha * y_history[i]
+    return r, alpha
```

Comment:
I guess you didn't need to handle pytrees in these helper functions after all?

Reply:
I'm not sure why but in the previous version of the code, I was getting errors if I didn't use tree utils, even though `inv_hessian_product_leaf` is supposed to only receive tree leaves... With the new version of the code, all tests passed directly with `jnp` functions so I went with that. Shall I use the tree utils instead?

Reply:
You have test coverage, so if things work with the new code, that's fine by me!